File: data-pipeline/archives/one_click_export.py
"""
一键导出功能

通过文件信号触发 REAPER 自动导出
"""

import os
import json
import time
from pathlib import Path
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

# 临时文件目录
TEMP_DIR = Path("/tmp/synest_export")
SIGNAL_FILE = TEMP_DIR / "trigger.txt"
CONFIG_FILE = TEMP_DIR / "config.json"
RESULT_FILE = TEMP_DIR / "result.json"


def trigger_reaper_export(project_name, theme_name, render_preview=True):
    """
    触发 REAPER 导出

    通过创建信号文件,通知 REAPER 开始导出
    """
    # 创建临时目录
    TEMP_DIR.mkdir(parents=True, exist_ok=True)

    # 准备配置
    config = {
        "project_name": project_name,
        "theme_name": theme_name,
        "render_preview": render_preview,
        "output_dir": str(Path(__file__).parent / "output")
    }

    # 写入配置文件
    with open(CONFIG_FILE, 'w', encoding='utf-8') as f:
        json.dump(config, f, indent=2, ensure_ascii=False)

    # 创建信号文件 (通知 REAPER)
    with open(SIGNAL_FILE, 'w') as f:
        f.write(str(CONFIG_FILE) + '\n')
        f.write(str(RESULT_FILE) + '\n')

    print(f"✓ 已触发 REAPER 导出")
    print(f"  项目: {project_name}")
    print(f"  主题: {theme_name}")
    print(f"  预览: {render_preview}")
    print(f"\n请在 REAPER 中按以下操作:")
    print(f"  1. 打开 Actions 列表 (Actions → Show action list)")
    print(f"  2. 搜索并运行: one_click_export")
    print(f"  3. 或手动运行: data-pipeline/lua_scripts/one_click_export.lua")

    return {
        'success': True,
        'signal_file': str(SIGNAL_FILE),
        'config_file': str(CONFIG_FILE)
    }


def wait_for_export_result(timeout=60):
    """
    等待 REAPER 导出完成

    监控结果文件,直到 REAPER 写入导出结果
    """
    print(f"\n等待 REAPER 导出完成...")

    start_time = time.time()
    while time.time() - start_time < timeout:
        if RESULT_FILE.exists():
            try:
                with open(RESULT_FILE, 'r', encoding='utf-8') as f:
                    result = json.load(f)

                # 清理临时文件
                SIGNAL_FILE.unlink(missing_ok=True)
                RESULT_FILE.unlink(missing_ok=True)

                return result
            except:
                pass

        time.sleep(1)

    return {
        'success': False,
        'error': f'等待超时 ({timeout}秒)'
    }


def clean_temp_files():
    """清理临时文件"""
    SIGNAL_FILE.unlink(missing_ok=True)
    CONFIG_FILE.unlink(missing_ok=True)
    RESULT_FILE.unlink(missing_ok=True)


if __name__ == '__main__':
    import sys

    try:
        if len(sys.argv) > 1:
            project_name = sys.argv[1]
            theme_name = sys.argv[2] if len(sys.argv) > 2 else "默认主题"
        else:
            project_name = input("项目名: ")
            theme_name = input("主题名: ")

        # 触发导出
        result = trigger_reaper_export(project_name, theme_name)

        if result['success']:
            # 等待完成
            final_result = wait_for_export_result()

            if final_result.get('success'):
                print(f"\n✓ 导出成功!")
                print(f"  胶囊ID: {final_result.get('capsule_id')}")
            else:
                print(f"\n✗ 导出失败: {final_result.get('error')}")

    except KeyboardInterrupt:
        print("\n\n用户取消")
        clean_temp_files()
        sys.exit(1)
    finally:
        clean_temp_files()

File: data-pipeline/lua_scripts/one_click_export.lua
-- 一键导出脚本
-- 功能: 检测信号文件并自动执行导出

local TEMP_DIR = "/tmp/synest_export"
local SIGNAL_FILE = TEMP_DIR .. "/trigger.txt"
local CONFIG_FILE = TEMP_DIR .. "/config.json"
local RESULT_FILE = TEMP_DIR .. "/result.json"

-- 检查信号文件
local function check_signal_file()
    local file = io.open(SIGNAL_FILE, "r")
    if file then
        file:close()
        return true
    end
    return false
end

-- 读取配置
local function read_config()
    local file = io.open(CONFIG_FILE, "r")
    if not file then
        return nil
    end

    local content = file:read("*all")
    file:close()

    return json.decode(content)
end

-- 写入结果
local function write_result(result)
    local file = io.open(RESULT_FILE, "w")
    if not file then
        reaper.ShowConsoleMsg("无法写入结果文件\n")
        return false
    end

    file:write(json.encode(result))
    file:close()

    return true
end

-- 主导出函数
local function do_export(config)
    reaper.ShowConsoleMsg("开始一键导出...\n")

    -- 获取选中的 Items
    local selected_items = {}
    local item_count = reaper.CountSelectedMediaItems(0)

    if item_count == 0 then
        return {
            success = false,
            error = "没有选中的 Item"
        }
    end

    for i = 0, item_count - 1 do
        local item = reaper.GetSelectedMediaItem(0, i)
        table.insert(selected_items, item)
    end

    reaper.ShowConsoleMsg(string.format("找到 %d 个选中的 Item\n", item_count))

    -- 这里应该调用实际的导出逻辑
    -- 为了简化,我们暂时返回成功

    local result = {
        success = true,
        capsule_name = config.project_name .. "_" .. config.theme_name,
        message = "导出成功"
    }

    return result
end

-- 主循环
function main()
    -- 检查信号文件
    if not check_signal_file() then
        return
    end

    reaper.ShowConsoleMsg("检测到导出信号\n")

    -- 读取配置
    local config = read_config()
    if not config then
        reaper.ShowConsoleMsg("无法读取配置文件\n")
        return
    end

    reaper.ShowConsoleMsg(string.format("项目: %s\n", config.project_name))
    reaper.ShowConsoleMsg(string.format("主题: %s\n", config.theme_name))

    -- 执行导出
    local result = do_export(config)

    -- 写入结果
    write_result(result)

    reaper.ShowConsoleMsg("导出完成\n")
end

-- 执行
main()
